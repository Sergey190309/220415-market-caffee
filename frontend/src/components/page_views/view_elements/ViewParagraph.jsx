import React, { useState, useEffect, useRef, Fragment } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import PropTypes from 'prop-types'
import { Message, Divider } from 'semantic-ui-react'

import {
  CONTENT_REQUESTED
} from '../../../redux/constants/types'
import { useSaga } from '../../../redux/saga/content/createIO'
import {
  getContentSaga
  // putContentSaga
} from '../../../redux/saga/content/content'
import {
  deviceSelector,
  backendUpdateSelector,
  backendTxtUpdateReady, backendTxtUpdateStart,
  resetBackendUpdate
} from '../../../redux/slices'
import { createContextFromEvent } from './editors/createContextFromEvent' // tested
import ParagraphContextMenu
  from './editors/ParagraphContextMenu' // tested
import SaveToBackendContextMenu from '../../items/SaveToBackendContextMenu'
import ParagraphEditor from './editors/ParagraphEditor' // tested
import Indicator from './indicator/Indicator'

const ViewParagraph = ({
  initialState, recordId, viewName,
  addElementProp, deleteElementProp
  // upperLvlAddElementProp, upperLvlDeleteElementProp
}) => {
  /**
   * States:
   * state: object - Paragraph content loaded from back-end
   *    using getContentSaga.
   * content: object - Content itself shown on the component.
   *    Updated with useEffect.
   * changed: boolean - Indication content was changed from last
   *    download or upload. Updated with useEffect.
   * contextMenuOpened: boolean - Self explain. Open with right
   *    button.
   * paragraphEditted: boolean - Set close showing, open
   *    textboxes for edition.
   * indicatorOpened: boolean - Set component indicator on and
   *    off respectevily.
   * Store state variables.
   * editable: boolean - Admin only can set this variable on
   *    admin page.
   * loaded: boolean - Indication whether changed content
   *    successfully uploaded to back-end. Used to set save to
   *    back-end context menu disabled.
   */
  const [state, getSagaDispatch] = useSaga(
    getContentSaga, initialState)
  const [content, setContent] = useState({
    title: '',
    content: ['']
  })
  const [changed, setChanged] = useState(false)
  const [contextMenuOpened, setContextMenuOpened] = useState(false)
  const [paragraphEditted, setParagraphEditted] = useState(false)
  const [indicatorOpened, setIndicatorOpened] = useState(false)
  const { editable } = useSelector(deviceSelector)
  const { loaded, kind } = useSelector(backendUpdateSelector)
  const dispatch = useDispatch()

  const contextRef = useRef(null)
  const indicatorRef = useRef(null)

  useEffect(() => { // Saga
    // console.log('ViewParagraph, useEffect(getSagaDispatch), recordId ->', recordId)
    getSagaDispatch({
      type: CONTENT_REQUESTED,
      payload: {
        identity: recordId,
        view_id: viewName
      }
    })
  }, [recordId, viewName])

  useEffect(() => {
    setContent(state)
  }, [state])

  useEffect(() => {
    if (JSON.stringify(state) !== JSON.stringify(content)) {
      // console.log('ViewParagraph:\n useEffect (content)',
      //   '\n  state ->', state,
      //   '\n  content', content
      // )
      dispatch(backendTxtUpdateReady({
        identity: recordId,
        view_id: viewName,
        content: content
      }))
      setChanged(true)
    }
  }, [content])

  useEffect(() => {
    if (changed) {
      // console.log('componetns, ViewParagraph, useEffect ->')
      setChanged(false)
      dispatch(resetBackendUpdate())
    }
  }, [loaded])

  const onClickHandler = event => {
    // console.log('components, page_view, view_elements, ViewParagraph, identities ->', identities)
    event.preventDefault()
    if (editable) {
      setIndicatorOpened(editable && !indicatorOpened)
      indicatorRef.current = createContextFromEvent(event)
    }
  }

  const onContextMenuHendler = event => {
    event.preventDefault()
    contextRef.current = createContextFromEvent(event)
    setIndicatorOpened(false)
    setContextMenuOpened(true)
  }

  const saveToBackend = () => {
    // console.log('ViewParagraph, saveToBackend, content ->', content)
    dispatch(backendTxtUpdateStart())
  }
  const addAbove = () => {
    /**
     * To send signal one module above to change structure
     */
    const recordIndex = parseInt(recordId.split('_').pop())
    if (isNaN(recordIndex)) {
      console.log('Unable to parse recordId for index')
    } else { addElementProp(recordIndex) }
  }
  const addBelow = () => {
    /**
     * To send signal one module above to change structure
     */
    const recordIndex = parseInt(recordId.split('_').pop())
    if (isNaN(recordIndex)) {
      console.log('Unable to parse recordId for index')
    } else { addElementProp(recordIndex + 1) }
  }
  const deleteElement = () => {
    /**
     * To send signal one module above to change structure
     */
    const recordIndex = parseInt(recordId.split('_').pop())
    // console.log('ViewParagraph, deleteFmBackend')
    if (isNaN(recordIndex)) {
      console.log('Unable to parse recordId for index')
    } else { deleteElementProp(recordIndex) }
  }

  const normalOutput = () => (
    <Message
      onContextMenu={editable
        ? onContextMenuHendler
        : null
      }
      onClick={onClickHandler}
      data-testid='Message'
    >
      <Message.Header content={content.title} />
      {content.title && content.content.length > 0
        ? <Divider />
        : null}
      {content.content.map((item, index) => (
        <Message.Item as='p' key={index}>
          {item}
        </Message.Item>
      ))}
    </Message>
  )

  const paragraphEditorConf = () => (
    <ParagraphEditor
      setParagraphEditted={setParagraphEditted}
      comingContent={content}
      setComingContent={setContent}
    />
  )

  const indicatorConf = () => (
    <Indicator
      isOpened={indicatorOpened}
      context={indicatorRef}
      header={viewName}
      content={recordId}
      setIndicatorOpened={setIndicatorOpened}
    />
  )

  const paragraphContextMenuConf = () => (
    <ParagraphContextMenu
      isOpened={contextMenuOpened}
      saveDisabled={!changed}
      context={contextRef}
      setContextMenuOpened={setContextMenuOpened}
      setParagraphEditted={setParagraphEditted}
      saveToBackend={saveToBackend}
      deleteElement={deleteElement}
      addAbove={addAbove}
      addBelow={addBelow}
    />
  )

  return (
    <Fragment>
      {editable
        ? paragraphEditted
          ? paragraphEditorConf()
          : <Fragment>
              {normalOutput()}
              {indicatorOpened
                ? indicatorConf()
                : null
              }
              {contextMenuOpened
                ? (kind === '')
                    ? paragraphContextMenuConf()
                    : changed
                      ? paragraphContextMenuConf()
                      : <SaveToBackendContextMenu />
                : null
              }
            </Fragment>
        : normalOutput()
      }
      {/* {editable
        ? paragraphEditted
          ? <ParagraphEditor
            setParagraphEditted={setParagraphEditted}
            comingContent={content}
            setComingContent={setContent}
          />
          : <ParagraphContextMenu
            isOpened={contextMenuOpened}
            // saveDisabled={false}
            saveDisabled={!changed}
            context={contextRef}
            setContextMenuOpened={setContextMenuOpened}
            setParagraphEditted={setParagraphEditted}
            saveToBackend={saveToBackend}
            deleteElement={deleteElement}
            addAbove={addAbove}
            addBelow={addBelow}
          // upperLvlAddElementProp={upperLvlAddElementProp}
          // upperLvlDeleteElementProp={upperLvlDeleteElementProp}
          />
        : null
      }
      {paragraphEditted
        ? null
        : <NormalOutput />
      }
      {indicatorOpened
        ? <Indicator
          isOpened={indicatorOpened}
          context={indicatorRef}
          header={viewName}
          content={recordId}
          setIndicatorOpened={setIndicatorOpened}
        />
        : null
      } */}
    </Fragment>
  )
}

ViewParagraph.defaultProps = {
  initialState: {
    title: '', // info from back-end
    content: ['']
  },
  recordId: '',
  viewName: '',
  addElementProp: () => { },
  deleteElementProp: () => { }
  // upperLvlAddElementProp: () => { },
  // upperLvlDeleteElementProp: () => {}
}

ViewParagraph.propTypes = {
  initialState: PropTypes.object.isRequired,
  recordId: PropTypes.string.isRequired,
  viewName: PropTypes.string.isRequired,
  addElementProp: PropTypes.func.isRequired,
  deleteElementProp: PropTypes.func.isRequired
  // upperLvlAddElementProp: PropTypes.func.isRequired,
  // upperLvlDeleteElementProp: PropTypes.func.isRequired
}

export default ViewParagraph
