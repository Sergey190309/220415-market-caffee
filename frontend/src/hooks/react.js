import {useState, useEffect, useRef, useCallback } from 'react'

export const useAppState = useState
export const useAppEffect = useEffect
export const useAppRef = useRef
export const useAppCallback = useCallback
