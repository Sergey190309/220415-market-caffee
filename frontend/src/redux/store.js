import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit'
import createSagaMiddleware from 'redux-saga'
// import reducer from './reducers';
import reducer from './slices'
import rootSaga from './saga'
// import { rootReducer } from './reducers';

const initialState = {}
const sagaMiddleware = createSagaMiddleware()

const middleware = [...getDefaultMiddleware({ thunk: false }), sagaMiddleware]
// const middleware = [sagaMiddleware, thunk];

const store = configureStore({
  reducer,
  middleware,
  initialState
})

sagaMiddleware.run(rootSaga)

export default store
