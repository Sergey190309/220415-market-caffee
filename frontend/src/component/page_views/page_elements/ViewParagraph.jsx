import React from 'react'
import { useAppState, useAppEffect, useAppContext } from '../../../hooks/react'
import { useAppSelector } from '../../../hooks/reactRedux'
import PropTypes from 'prop-types'

import { deviceSelector } from '../../../redux/slices'
import { CONTENT_REQUESTED } from '../../../redux/constants/types'
import { useSaga } from '../../../redux/contentSaga/createIO'
import { getContentSaga } from '../../../redux/contentSaga/content'
import { LandingContext } from '../../../context'
import { Box, Divider, Tooltip, Typography } from '@mui/material'
import * as SZ from '../../../constants/sizes'
import * as CL from '../../../constants/colors'

const ViewParagraph = ({ recordId, initialState }) => {
  /**
  * States:
  * state: object - Paragraph content loaded from back-end
  *    using getContentSaga.
  * content: object - Content itself shown on the component.
  *    Updated with useEffect.
  * changed: boolean - Indication content was changed from last
  *    download or upload. Updated with useEffect.
  * contextMenuOpened: boolean - Self explain. Open with right
  *    button.
  * upperLevelContextMenuOpened: boolean. Open when appropriate button
  *    pressed on context menu.
  * paragraphEditted: boolean - Set close showing, open
  *    textboxes for edition.
  * indicatorOpened: boolean - Set component indicator on and
  *    off respectevily.
  */
  const [content, setContent] = useAppState(initialState)
  const [changed, setChanged] = useAppState(false)
  const [tooltipVisible, setTooltipVisible] = useAppState(false)

  const [state, getSagaDispatch] = useSaga(getContentSaga, initialState)

  const { editable } = useAppSelector(deviceSelector)

  const { componentName } = useAppContext(LandingContext)
  const { componentName: viewName } = useAppContext(LandingContext)

  useAppEffect(() => { // Saga
    // console.log('ViewParagraph:',
    //   '\n useEffect[recordId, kind]',
    //   '\n  recordId ->', recordId)
    // indexRef.current = +recordId.split('_').pop()
    // if (kind === '') {
      setChanged(false)
      getSagaDispatch({
        type: CONTENT_REQUESTED,
        payload: {
          identity: recordId,
          view_id: viewName
        }
      })
    // setContent(state)
    // console.log('ViewParagraph, state ->', state)
    // }
    // }, [recordId, kind])
  }, [])

  useAppEffect(() => {
    setContent(state)
  }, [state])


  const NormalOutput = () => (
    <Tooltip
      title={recordId} placement='top' followCursor arrow
      open={tooltipVisible}
      onOpen={() => { setTooltipVisible(editable && true) }}
      onClose={() => { setTooltipVisible(false) }}
    >
      <Box
        sx={{
          border: SZ.blockBorder, borderColor: 'text.disabled', borderRadius: 3,
          m: '.25rem',
          p: '.5rem',
          '&:hover': editable && {
            borderColor: CL.attention
          }
        }}
      >
        {/* Fuck! */}
        <Typography variant='h6'>
          {content.title}
        </Typography>
        {content.title && content.content.length > 0
          ? <Divider />
          : null}
        {content.content.map((item, index) => (
          <Typography key={index} variant='body2'>
            {item}
          </Typography>
        ))}
      </Box>
    </Tooltip>
  )

  return (
    <NormalOutput />
  )
}
ViewParagraph.defaultProps = {
  recordId: '',
  initialState: {
    title: '', // info from back-end
    content: ['']
  }
}
ViewParagraph.propTypes = {
  recordId: PropTypes.string.isRequired,
  initialState: PropTypes.object.isRequired
}

export default ViewParagraph