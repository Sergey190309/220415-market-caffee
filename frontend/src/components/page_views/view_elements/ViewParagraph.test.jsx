/* eslint-disable react/display-name */
import React from 'react'
import { Provider } from 'react-redux'
import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import ViewParagraph from './ViewParagraph'
import store from '../../../redux/store'
import { resolveDataContent } from '../../../testAxiosConstants'

jest.mock('../../../api/calls/content')

const mockParagraphContextMenu = jest.fn()

jest.mock('./editors/ParagraphContextMenu', () => props => {
  mockParagraphContextMenu(props)
  return <div data-testid='mockParagraphContextMenu' />
})

describe('ViewParagraph testing', () => {
  const testProps = {
    recordId: 'mockRecordId',
    viewName: 'mockViewName',
    lng: 'mockLng'
  }
  describe('actions', () => {
    test('right click -> context menu with appropriate args', async () => {
      await waitFor(() => {
        render(
          <Provider store={store}>
            <ViewParagraph {...testProps} />
          </Provider>
        )
      })
      const element = screen.getByTestId('Message')
      userEvent.click(element, { button: 2 })
      const result = mockParagraphContextMenu.mock
        .calls[0][0]
      expect(result.saveDisabled).toBe(true)
      expect(result.context).toBeObject()
      expect(result.setContextMenuOpened).toBeFunction()
      expect(result.setParagraphEditted).toBeFunction()
      expect(result.saveToBackend).toBeFunction()
      expect(result.deleteElement).toBeFunction()
      expect(result.addAbove).toBeFunction()
      expect(result.addBelow).toBeFunction()
      // console.log('ViewParagraph.test, right click, calls ->', result)
    })

    test('should indicate proper values', async () => {
      await waitFor(() => {
        render(
          <Provider store={store}>
            <ViewParagraph {...testProps} />
          </Provider>
        )
      })
      const result = screen.getByTestId('Message')
      expect(result.children).toHaveLength(5)
      expect(result.children[0]).toHaveTextContent(resolveDataContent.payload.title)
      expect(result.children[1]).toHaveClass('ui divider')

      expect(result.children[2]).toHaveTextContent(resolveDataContent.payload.content[0])
      expect(result.children[3]).toHaveTextContent(resolveDataContent.payload.content[1])
      expect(result.children[4]).toHaveTextContent(resolveDataContent.payload.content[2])
      // screen.debug()
    })
    test('should indicate proper values, no title', async () => {
      const actualProps = {
        ...testProps,
        recordId: 'mockNoTitle'
      }
      await waitFor(() => {
        render(
          <Provider store={store}>
            <ViewParagraph {...actualProps} />
          </Provider>
        )
      })
      const result = screen.getByTestId('Message')
      expect(result.children).toHaveLength(4)
      // expect(result.children[0]).toHaveTextContent(resolveDataContent.payload.title)
      // expect(result.children[1]).toHaveClass('ui divider')

      expect(result.children[0]).toBeEmptyDOMElement()
      expect(result.children[1]).toHaveTextContent(resolveDataContent.payload.content[0])
      expect(result.children[2]).toHaveTextContent(resolveDataContent.payload.content[1])
      expect(result.children[3]).toHaveTextContent(resolveDataContent.payload.content[2])
      // screen.debug()
    })

    test('should indicate proper values, no content', async () => {
      const actualProps = {
        ...testProps,
        recordId: 'mockNoContent'
      }
      await waitFor(() => {
        render(
          <Provider store={store}>
            <ViewParagraph {...actualProps} />
          </Provider>
        )
      })
      const result = screen.getByTestId('Message')
      expect(result.children).toHaveLength(1)
      expect(result.children[0]).toHaveTextContent(resolveDataContent.payload.title)
      // screen.debug()
    })
  })

  describe('appeance', () => {
    test('rendering (shapshot)', async () => {
      await waitFor(() => {
        render(
          <Provider store={store}>
            <ViewParagraph {...testProps} />
          </Provider>
        )
      })
      const result = screen.getByTestId('Message')
      // console.log('ViewParagraph testing, appeance, result ->', result)
      expect(result).toMatchSnapshot()
      // screen.debug()
    })
  })
})
