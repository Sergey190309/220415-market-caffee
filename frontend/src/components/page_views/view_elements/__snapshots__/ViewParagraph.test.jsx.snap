// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewParagraph testing appearance rendering normal (shapshot) 1`] = `
<div
  class="ui message"
  data-testid="Message"
>
  <div
    class="header"
  >
    mockTitle
  </div>
  <div
    class="ui divider"
  />
  <p
    class="content"
  >
    mock content first line
  </p>
  <p
    class="content"
  >
    mock content second line
  </p>
  <p
    class="content"
  >
    mock content third line
  </p>
</div>
`;
